gobblin-core/src/main/java/org/apache/gobblin/source/extractor/extract/restapi/RestApiConnector.java

```diff
@@ -84,6 +85,10 @@ public RestApiConnector(State state) {
   public void close() throws IOException {
     // This is to close any idle connections opening by the httpClient
     this.closer.close();
+    if (this.getHttpClient() != null && !(this.getHttpClient() instanceof Closeable)) {
+      log.warn("httpClient is not closable, we will only close the idle connections");
+      this.getHttpClient().getConnectionManager().closeIdleConnections(0, TimeUnit.MILLISECONDS);
```

Comment:
you could close only the idle ones here, now... but why not make a shim impl of `Closeable` that you could wrap it in and register w/ the `Closer`?  then, later, when that `Closer.close()` is called, you could handle all the CM cnxns, idle or not.  the difference in that approach is both closing all cnxns--but also that the time when doing the close would be later, such that more idle cnxns should develop.

stepping back, the seriousness of whether 'enough' to log and close only idle versus to be certain to close all CM cnxns would seem to come down to how prevalent is our use of `non-Closeable` `HttpClient`s.  strictly speaking however it looks like a resource leak not to close all of them always, by some point, even if the situation of missing to close some be rare.  do you agree?

Reply:
sorry, I misunderstood where you inserted these lines, when I viewed the limited diff of 'new changes' a moment ago.  my comment may not make sense.... let me revise:

now seems you're doing the same RTTI check in two places.  if it passes the first time, you register w/ the `Closer` that's used here.  otherwise you follow up on the negation of the same check to take action here.  that does handle both run-time paths I was concerned about (modulo the connections not yet 'idle').

1. are we concerned about non-idle connections?
2. I have slight preference for wrapping in a `Closer` impl (as described in prior comment) which does the RTTI once--not for efficiency, but to encapsulate in a singular place.  your choice.

Reply:
I see.  I'm not super familar with best practices for non-closeable clients, and so presumed the way to safely handle all resources would be `ClientConnectionManager.shutdown()`.  that is what my recommendation was based on.  but if that's not the way to full resource correctness, then definitely seems to be a gap in our ability to fully account for resources... and I wouldn't have any suggestions beyond what you have here.